p2p/protocol/autonatv2/server.go

```diff
@@ -0,0 +1,410 @@
+		return
+	}
+
+	nonce := msg.GetDialRequest().Nonce
+	// parse peer's addresses
+	var dialAddr ma.Multiaddr
+	var addrIdx int
+	for i, ab := range msg.GetDialRequest().GetAddrs() {
+		if i >= maxPeerAddresses {
+			break
+		}
+		a, err := ma.NewMultiaddrBytes(ab)
+		if err != nil {
+			continue
+		}
+		if !as.allowPrivateAddrs && !manet.IsPublicAddr(a) {
+			continue
+		}
+		if _, err := a.ValueForProtocol(ma.P_CIRCUIT); err == nil {
+			continue
+		}
+		if as.dialerHost.Network().CanDial(p, a) != network.DialabilityDialable {
+			continue
+		}
+		// Check if the host can dial the address. This check ensures that we do not
+		// attempt dialing an IPv6 address if we have no IPv6 connectivity as the host's
+		// black hole detector is likely to be more accurate.
+		if as.host.Network().CanDial(p, a) != network.DialabilityDialable {
+			continue
+		}
+		dialAddr = a
+		addrIdx = i
+		break
+	}
+
+	// No dialable address
+	if dialAddr == nil {
+		msg = pb.Message{
+			Msg: &pb.Message_DialResponse{
+				DialResponse: &pb.DialResponse{
+					Status: pb.DialResponse_E_DIAL_REFUSED,
+				},
+			},
+		}
+		if err := w.WriteMsg(&msg); err != nil {
+			s.Reset()
+			log.Debugf("failed to write dial refused response to %s: %s", p, err)
+			return
+		}
+		return
+	}
+
+	isDialDataRequired := as.dialDataRequestPolicy(s, dialAddr)
+	if !as.limiter.AcceptDialDataRequest(p) {
+		msg = pb.Message{
+			Msg: &pb.Message_DialResponse{
+				DialResponse: &pb.DialResponse{
+					Status: pb.DialResponse_E_REQUEST_REJECTED,
+				},
+			},
+		}
+		if err := w.WriteMsg(&msg); err != nil {
+			s.Reset()
+			log.Debugf("failed to write request rejected response to %s: %s", p, err)
+			return
+		}
+		log.Debugf("rejected request from %s: rate limit exceeded", p)
+		return
+	}
+
+	if isDialDataRequired {
+		if err := getDialData(w, r, &msg, addrIdx); err != nil {
+			s.Reset()
+			log.Debugf("%s refused dial data request: %s", p, err)
+			return
+		}
+	}
+
+	dialStatus := as.dialBack(s.Conn().RemotePeer(), dialAddr, nonce)
+	msg = pb.Message{
+		Msg: &pb.Message_DialResponse{
+			DialResponse: &pb.DialResponse{
+				Status:     pb.DialResponse_OK,
+				DialStatus: dialStatus,
+				AddrIdx:    uint32(addrIdx),
+			},
+		},
+	}
+	if err := w.WriteMsg(&msg); err != nil {
+		s.Reset()
+		log.Debugf("failed to write response to %s: %s", p, err)
+		return
+	}
+}
+
+// getDialData gets data from the client for dialing the address
+func getDialData(w pbio.Writer, r pbio.Reader, msg *pb.Message, addrIdx int) error {
+	numBytes := minHandshakeSizeBytes + rand.Intn(maxHandshakeSizeBytes-minHandshakeSizeBytes)
+	*msg = pb.Message{
+		Msg: &pb.Message_DialDataRequest{
+			DialDataRequest: &pb.DialDataRequest{
+				AddrIdx:  uint32(addrIdx),
+				NumBytes: uint64(numBytes),
+			},
+		},
+	}
+	if err := w.WriteMsg(msg); err != nil {
+		return fmt.Errorf("dial data write: %w", err)
+	}
+	for remain := numBytes; remain > 0; {
+		if err := r.ReadMsg(msg); err != nil {
```

Comment:
Unfortunately I think this will allocate on every loop iteration. Example: https://gist.github.com/MarcoPolo/37338f09fe97f3ce7300f0397da76e30

An alternative would be to just read expected num of bytes from the stream not wrapped it in a pb.

Reply:
We might need to parse the protobuf because in some cases we would be leaving 1 message unread in the stream(data request size 4k + 1). This is fine for now, but I prefer to not have it. 

We can also change the spec to send raw bytes for `DialDataRequest`. In case we add other charge types for dialing an address besides dial data we can support those without any breaking change.

Reply:
I think changing the spec to just send and expect empty bytes is fine. I don't think we gain anything by wrapping it in a pb.

Reply:
Rust libp2p implementation prefers to have typed streams. I believe the idea is that you can wrap streams like: `Stream<AutoNATMsg>` so all read calls return object of the type `AutoNATMsg`

Reply:
I did custom protobuf parsing. Now the number of allocations are 4 for send side and 2 for read side for the entire dial data(100kB).

Reply:
What were they before?

Reply:
Answering my question, I think around 10 times more.